replication/AmbryReplication.tla

```diff
@@ -0,0 +1,328 @@
+        /\ [blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r1]
+        /\ storedMessagesPerReplica' = [rr \in DOMAIN storedMessagesPerReplica |-> 
+            IF ((rr = r1) \/ (rr = r2)) THEN
+                storedMessagesPerReplica[rr] \union { [ blobid |-> blobId, blobmessage |-> DELETE ] }
+            ELSE
+                storedMessagesPerReplica[rr]
+            ]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, updateTtlCtr, inflightReplicationRequest, inflightReplicationResponse, recvReplicationReqSet, sendRecplicationSet, recvReplicationRespSet>>
+
+(* Receive a UPDATE TTL request with success target of 2.
+ * For the request to succeed, atleast two replicas
+ *    - must be up AND
+ *    - must have a PUT record for the blob AND
+ *    - must not have a DELETE record for the blob.
+ *)
+RecvUpdateTtl(blobId) ==
+    /\ \E r1, r2 \in Replicas:
+        /\ r1 # r2
+        /\ blobId < blobIdCtr
+        /\ replicaStates[r1] = CanTakeWrite
+        /\ replicaStates[r2] = CanTakeWrite
+        /\ [ blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r1]
+        /\ [ blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r1]
+        /\ [ blobid |-> blobId, blobmessage |-> PUT ] \in storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> DELETE ] \notin storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] \notin storedMessagesPerReplica[r2]
+        /\ [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] \notin storedMessagesPerReplica[r1]
+        /\ storedMessagesPerReplica' = [rr \in DOMAIN storedMessagesPerReplica |-> 
+            IF ((rr = r1) \/ (rr = r2)) THEN
+                storedMessagesPerReplica[rr] \union { [ blobid |-> blobId, blobmessage |-> UPDATE_TTL ] }
+            ELSE
+                storedMessagesPerReplica[rr]
+            ]
+        /\ updateTtlCtr' = updateTtlCtr + 1
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, deleteCtr, inflightReplicationRequest, inflightReplicationResponse, recvReplicationReqSet, sendRecplicationSet, recvReplicationRespSet>>
+
+(* Bring one replica up and one replica down *)
+ReplicaStateChange ==
+    \E m, n \in Replicas:
+        \/ ReplicaDown(n)
+        \/ ReplicaUp(m)
+
+(* Ambry uses a pull model for replication.
+ * In SendReplicationRequest action, a replica asks one of its peers to give the set of data the peer has.
+ * This action just enqueues the request to peer in "inflightReplicationRequest".
+ *)
+SendReplicationRequest(src) ==
+    /\ replicaStates[src] = CanTakeWrite \* We want replica asking for replication pull to be writable.
+    /\ \E tgt \in Replicas:
+        /\ src # tgt
+        /\ inflightReplicationRequest' = [inflightReplicationRequest EXCEPT ![tgt] = inflightReplicationRequest[tgt] \union {src} ]
+    /\ IF AllBlobOpsDone THEN 
```

Comment:
I might need some elaboration on this.  If all operations are done when sendReplicationRequest is triggered, we're going to add the src to the sendReplicationSet, and not touch it otherwise?  This runs opposite to what I'd assume from the parsing of AllBlobOpsDone (to me at least).  Why is this?